Add a coordinator to Waze Travel Time

etiennec78 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR only moves the data retrieval from the sensor to a new coordinator.
I wanted to make the least amount of modifications for now. I will use a proper dataclass in the following PR.

Next PRs planned for Waze Travel Time:
1. Catch errors with UpdateFailed to avoid reporting old states as new
2. Use a proper dataclass and merge fetching functions
3. Add a custom available function to avoid reporting "unknown"
4. Remove rounding on the state
5. Split attributes into distinct sensors


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #146798
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

etiennec78 committed:
Add a coordinator to Waze Travel Time

etiennec78 committed:
Fix condition

etiennec78 committed:
Use a more standard function for value

eifinger reviewed (approved):
Looks good!

etiennec78 commented on homeassistant/components/waze_travel_time/sensor.py:
```diff
@@ -2,56 +2,25 @@
-PARALLEL_UPDATES = 1
-
-SECONDS_BETWEEN_API_CALLS = 0.5
+_LOGGER = logging.getLogger(__name__)
```
Just to be sure, you're only talking about SECONDS_BETWEEN_API_CALLS ?

eifinger replied:
No. Only about the logger. I think its not used anymore

etiennec78 committed:
Remove logger in sensor

joostlek commented on homeassistant/components/waze_travel_time/coordinator.py:
```diff
@@ -0,0 +1,239 @@
+        return filtered_routes
+
+
+class WazeTravelTimeData:
+    """WazeTravelTime Data object."""
+
+    def __init__(
+        self, region: str, client: httpx.AsyncClient, config_entry: ConfigEntry
+    ) -> None:
+        """Set up WazeRouteCalculator."""
+        self.config_entry = config_entry
+        self.client = WazeRouteCalculator(region=region, client=client)
+        self.origin: str | None = None
+        self.destination: str | None = None
+        self.duration = None
+        self.distance = None
+        self.route = None
+
+    async def async_update(self):
+        """Update WazeRouteCalculator data."""
```
I'm wondering what the benefit of this object is and why don't we just move the async_update method to the coordinator and return the result

etiennec78 replied:
I was asked to split my PR into smaller ones, so I thought it was a good idea to keep changes minimal in the first one.
You can check the description of this PR for the list of incoming PRs.
But don't hesitate to tell me if you want me to make more changes in this one.

etiennec78 committed:
Raise UpdateFailed instead of returning None

etiennec78 committed:
Remove redundant check

etiennec78 committed:
Merge fetching functions and use a proper dataclass

etiennec78 committed:
Updated tests

etiennec78 committed:
Fix typo in tests

It appears that a test contained a typo, but was not being triggered.
The route returned contained "IncludeThis" but the test was expecting "include".
Since raising UpdateFailed, this test now triggers and fails.
To fix this, all "include" strings have been replaced by "IncludeThis".

etiennec78 committed:
Log warning when routes list is empty

Log warning instead of raising UpdateFailed when no routes were found or no routes matched filters
